File: src/planner/planner.ts
import {Query} from "../parser";

export function plan(query: Query): Plan {
    const producerStep: ProducerStep = {
        name: "producer",
        table: query.source,
        columns: query.columns,
    };
    const filterStep: FilterStep = {
        name: "filter",
        columns: query.columns,
        property: query.filter[0],
        value: query.filter[1],
    };
    return [
        producerStep,
        filterStep
    ];
}

export type Plan = Step[];

export type Step = ProducerStep | FilterStep;

export type ProducerStep = {
    name: "producer",
    table: string,
    columns: string[],
}

export type FilterStep = {
    name: "filter",
    columns: string[],
    property: string,
    value: any,
}

File: rust/src/query/planner.rs
use serde_json::Value;

use crate::query::parser::Query;

pub struct Plan {
    pub producer: ProducerStep,
    pub filter: Option<FilterStep>,
}

pub struct ProducerStep {
    pub source: String,
    pub columns: Vec<String>,
}

pub struct FilterStep {
    columns: Vec<String>,
    property_name: String,
    value: Value,
}

pub fn plan(query: Query) -> Plan {
    Plan {
        producer: ProducerStep {
            source: query.source,
            columns: query.columns.clone(),
        },
        filter: query.filter.map(|predicate| FilterStep {
            columns: query.columns,
            property_name: predicate.property_name,
            value: predicate.value,
        })
    }
}
